InformationScripting/src/queries/CompositeQuery.cpp

```diff
@@ -0,0 +1,202 @@
+	QQueue<QueryNode*> workingSetNodes;
+	workingSetNodes.enqueue(inNode_);
+	QSet<QueryNode*> visitedNodes;
+
+	// Check if we have nodes which take no input:
+	for (auto queryNode : nodes_)
+	{
+		if (queryNode->inputs_.empty())
+		{
+			Q_ASSERT(queryNode->q_);
+			queryNode->execute();
+			workingSetNodes.enqueue(queryNode);
+		}
+	}
+
+	while (!workingSetNodes.empty())
+	{
+		auto currentNode = workingSetNodes.dequeue();
+		// TODO: once we allow cyclic dependencies we have to be smarter here:
+		if (visitedNodes.contains(currentNode)) continue;
+		visitedNodes.insert(currentNode);
+
+		for (auto outputMapping : currentNode->outputs_)
+		{
+			int outIndex = outputMapping.outputIndex_;
+			// TODO: this shouldn't be an assertion later, but rather some error which is presented to the user.
+			Q_ASSERT(outIndex < currentNode->calculatedOutputs_.size());
+
+			// FIXME: multiple output receivers requires Graph copy constructor:
+			Q_ASSERT(outputMapping.outputReceivers_.size() <= 1);
+
+			if (outputMapping.outputReceivers_.size() == 1)
+			{
+				auto receiver = *outputMapping.outputReceivers_.begin();
+				// find the input mapping of the receiver:
+				auto inputIt = std::find_if(receiver->inputs_.begin(), receiver->inputs_.end(),
+					[currentNode, outIndex] (const InputMapping& m) {
+					return m.outputFrom_ == currentNode && m.outputIndex_ == outIndex;
+				});
+				Q_ASSERT(inputIt != receiver->inputs_.end());
+				auto inputMapping = *inputIt;
+				receiver->addCalculatedInput(inputMapping.inputIndex_, currentNode->calculatedOutputs_[outIndex]);
+
+				if (receiver->canExecute())
+				{
+					receiver->execute();
+					workingSetNodes.enqueue(receiver);
+				}
+			}
+		}
+	}
+	return outNode_->calculatedOutputs_;
+}
+
+void CompositeQuery::addQuery(Query* q)
+{
+	if (!nodeForQuery(q)) nodes_.push_back(new QueryNode(q));
+}
+
+void CompositeQuery::connectQuery(Query* from, Query* to)
+{
+	connectQuery(from, 0, to, 0);
+}
+
+void CompositeQuery::connectQuery(Query* from, int outIndex, Query* to, int inIndex)
+{
+	auto fromNode = nodeForQuery(from);
+	auto toNode = nodeForQuery(to);
+
+	addOutputMapping(fromNode, outIndex, toNode);
+	addInputMapping(fromNode, outIndex, toNode, inIndex);
+}
+
+void CompositeQuery::connectToOutput(Query* from, int outIndex)
+{
+	auto fromNode = nodeForQuery(from);
+
+	addOutputMapping(fromNode, 0, outNode_);
+	addInputMapping(fromNode, 0, outNode_, outIndex);
+}
+
+CompositeQuery::QueryNode* CompositeQuery::nodeForQuery(Query* q)
+{
+	auto it = std::find_if(nodes_.begin(), nodes_.end(), [q](QueryNode* d) {return d->q_ == q;});
+	if (it != nodes_.end()) return *it;
+	auto newNode = new QueryNode(q);
+	nodes_.push_back(newNode);
+	return newNode;
+}
+
+void CompositeQuery::addOutputMapping(CompositeQuery::QueryNode* outNode, int outIndex,
+												  CompositeQuery::QueryNode* inNode)
+{
+	// Add output mapping
+	auto it = std::find_if(outNode->outputs_.begin(), outNode->outputs_.end(),
```

Comment:
wow, can't we just do something along the lines of `outNode->output(2)` ? The current code looks like total overkill.